contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,654 @@
+
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice() error
+       SetUpDeviceAt(dir string) error
+       GetVolumePath() string
+ }
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+ }
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+`/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/`
+
+- Volume mount path
+`/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/`
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link.
+- Kubelet creates a new symbolic link under the new global mount path when volume is attached to a Pod. Number of symbolic links
+is equal to the number of Pods which uses the same volume. Kubelet needs to manage both creation and deletion of symbolic links
+under the global mount path.
+ 
+```
+Global mount path for "Block" mode volume
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/
+```
+ 
+- Plugin creates a symbolic link under the new volume mount path. This symbolic link is not used to manage volume attach/detach
++status but is needed to keep compatibility of current scheme.
+
+```
+Volume mount path for "Block" mode volume
+/var/lib/kubelet/pods/{podUID}/volumeDevices/{escapeQualifiedPluginName}/{volumeName}/
+```
+ 
+# Volume binding matrix for statically provisioned volumes:
+
+| PV volumeMode | PVC volumeMode  | Result           |
+| --------------|:---------------:| ----------------:|
+|   unspecified | unspecified     | BIND             |
+|   unspecified | Block           | NO BIND          |
+|   unspecified | Filesystem      | BIND             |
+|   Block       | unspecified     | NO BIND          |
+|   Block       | Block           | BIND             |
+|   Block       | Filesystem      | NO BIND          |
+|   Filesystem  | Filesystem      | BIND             |
+|   Filesystem  | Block           | NO BIND          |
+|   Filesystem  | unspecified     | BIND             |
+
+
+
+* unspecified defaults to 'file/ext4' today for backwards compatibility and in mount_linux.go  
+
+
```

Comment:
There should be a section discussing what kinds of operations can be done with and without privileges.  I feel it is a common enough question that gets asked.

Reply:
The container runtime doesn't require privilege for the device change. So we aren't changing anything over how it works today as far as privileges go

Reply:
I mean the pod-level privileges.  Your pod still has to be privileged if it needs to mount inside the container.

Reply:
To make clear, when we pass block device to container runtime such as docker,
- operations such as dd, fdisk, mkfs inside the container doesn't require privilege
- but, mount operation inside the container requires 'privileged' or cap SYS_ADMIN

Reply:
@mtanino that is correct, but we don't need to change mounting privileges for block as this is true for all devices

Reply:
@erinboyd if someone partitions and creates a fs on a block device how would they use it w/o mounting somehow?

Reply:
@childsb this is how almost all filesystems work today, is it's a block and a plugin at mount time applies the filesystem. I don't see how this is different. @msau42 ?

Reply:
The difference is that before, kubelet was the one formatting and mounting the block device, and kubelet has all the privileges necessary.  Now we're opening up that capability to the user, so I think it is important to document somewhere exactly what permissions the user needs, even if it is not any different than what the runtime supports.

Also because in the goals, we state that you can use block storage as non-privileged, which is mostly true, except if you need to mount.

Reply:
We should state that the use case "as a user i want to create a partition and filesystem on a block device then mount it within the container" wont work for unprivileged containers.

Reply:
@childsb We removed this use case as we aren't addressing it in this proposal. We are addressing it in the next phase as stated in the goals section.